File: 2nd year/Laboratories/MAP/L9/CRUD/UI/Tui/MarkTui.cs
using System;
using System.Collections.Generic;
using CRUD.Domain;
using CRUD.Service;
using CRUD.Utils;

namespace CRUD.UI.Tui
{
    public class MarkTui : BaseTui
    {
        private readonly MarkService _service;

        public MarkTui(MarkService service)
        {
            _service = service;
            Initialize("Mark TUI", new Dictionary<string, Action>
            {
                {"Create mark", CreateMark},
                {"Read one mark", ReadOneMark},
                {"Read all marks", ReadAllMarks}
            });
        }

        private void CreateMark()
        {
            var mark = Read("TaskId", "StudentId", "Value", "Feedback", "Week");
            var didCreateMark = _service.CreateMark(mark);
            if (didCreateMark)
            {
                var maximumMarkValue = _service.GetMaximumMarkValue(mark);
                if (double.Parse(mark["Value"]) > maximumMarkValue)
                {
                    Write($"Mark value was set to {maximumMarkValue} because of the delay");
                }
            }

            Write(GenerateMessage(didCreateMark));
        }

        private void ReadOneMark()
        {
            var taskId = Read("TaskId");
            var studentId = Read("StudentId");
            var mark = _service.ReadOneMark(taskId, studentId);
            Write(mark == null
                ? "Mark not found..."
                : EntityUtils.DumpString<Mark, string>(mark));
        }

        private void ReadAllMarks()
        {
            foreach (var mark in _service.ReadAllMarks())
            {
                Write(EntityUtils.DumpString<Mark, string>(mark));
            }
        }
    }
}

File: 2nd year/Laboratories/MAP/7/src/main/java/ui/tui/MarkTUI.java
package ui.tui;

import service.MarkService;

import java.util.HashMap;
import java.util.Map;

class MarkTUI extends BaseTUI {
    private MarkService markService;

    MarkTUI(MarkService markService) {
        this.markService = markService;
        generateTUI(
                "Mark TUI",
                new HashMap<String, Runnable>() {{
                    put("Add mark", MarkTUI.this::addMark);
                }}
        );
    }

    private void addMark() {
        Map<String, String> markMap = readMap("taskId", "studentId", "value", "feedback");
        Boolean didCreateMark = markService.createMark(markMap);
        if (didCreateMark) {
            Double maximumMarkValue = markService.getMaximumMarkValue(markMap);
            if (Double.valueOf(markMap.get("value")) > maximumMarkValue) {
                System.console().printf("Mark value was set to %.2f because the task was due\n", maximumMarkValue);
            }
            System.console().printf("Saving mark to student's file %s\n",
                    markService.dumpMarkToStudentFile(markMap) ? "successful" : "failed");
        }
        System.console().printf("Operation %s\n", didCreateMark ? "successful" : "failed");
    }
}
